core/arch/riscv/mm/core_mmu_arch.c

```diff
@@ -0,0 +1,717 @@
+}
+
+void core_mmu_get_entry_primitive(const void *table, size_t level,
+				  size_t idx, paddr_t *pa, uint32_t *attr)
+{
+	struct mmu_pgt *pgt = (struct mmu_pgt *)table;
+	struct mmu_pte *pte = core_mmu_table_get_entry(pgt, idx);
+
+	if (core_mmu_entry_is_valid(pte)) {
+		if (pa)
+			*pa = pte_to_pa(pte);
+		if (attr)
+			*attr = pte_to_mattr(level, pte);
+	} else {
+		if (pa)
+			*pa = 0;
+		if (attr)
+			*attr = 0;
+	}
+}
+
+void core_mmu_set_entry_primitive(void *table, size_t level, size_t idx,
+				  paddr_t pa, uint32_t attr)
+{
+	struct mmu_pgt *pgt = (struct mmu_pgt *)table;
+	struct mmu_pte *pte = core_mmu_table_get_entry(pgt, idx);
+	uint8_t perms = mattr_to_perms(level, attr);
+
+	core_mmu_entry_set(pte, core_mmu_pte_create(pa_to_ppn(pa), perms));
+}
+
+static void set_user_va_idx(struct mmu_partition *prtn)
+{
+	struct mmu_pgt *pgt = NULL;
+	struct mmu_pte *pte = NULL;
+	unsigned int idx = 0;
+
+	pgt = core_mmu_get_root_pgt_va(prtn);
+
+	for (idx = 1 ; idx < RISCV_PTES_PER_PT; idx++) {
+		pte = core_mmu_table_get_entry(pgt, idx);
+		if (core_mmu_entry_is_invalid(pte)) {
+			user_va_idx = idx;
+			return;
+		}
+	}
+	if (user_va_idx < 0)
+		panic();
+}
+
+static struct mmu_pte *
+core_mmu_get_user_mapping_entry(struct mmu_partition *prtn)
+{
+	struct mmu_pgt *pgt = core_mmu_get_root_pgt_va(prtn);
+
+	assert(user_va_idx != -1);
+
+	return core_mmu_table_get_entry(pgt, user_va_idx);
+}
+
+void core_mmu_set_user_map(struct core_mmu_user_map *map)
+{
+	unsigned long satp = 0;
+	uint32_t exceptions = thread_mask_exceptions(THREAD_EXCP_ALL);
+	struct mmu_partition *prtn = core_mmu_get_prtn();
+	struct mmu_pte *pte = NULL;
+	unsigned long ptp = 0;
+
+	satp = read_satp();
+	/* Clear ASID */
+	satp &= ~SHIFT_U64(RISCV_SATP_ASID_MASK, RISCV_SATP_ASID_SHIFT);
+	pte = core_mmu_get_user_mapping_entry(prtn);
+	if (map && map->user_map) {
+		ptp = core_mmu_ptp_create(pa_to_ppn((paddr_t)map->user_map));
+		core_mmu_entry_set(pte, ptp);
+		core_mmu_table_write_barrier();
+		satp |= SHIFT_U64(map->asid, RISCV_SATP_ASID_SHIFT);
+		write_satp(satp);
+	} else {
+		core_mmu_entry_set(pte, 0);
+		core_mmu_table_write_barrier();
+	}
+
+	tlbi_all();
+	thread_unmask_exceptions(exceptions);
+}
+
+void core_mmu_get_user_va_range(vaddr_t *base, size_t *size)
+{
+	assert(user_va_idx != -1);
+
+	if (base)
+		*base =  (vaddr_t)SHIFT_U64(user_va_idx,
+					    CORE_MMU_USER_TABLE_SHIFT);
+
+	if (size)
+		*size =  BIT64(CORE_MMU_USER_TABLE_SHIFT);
+}
+
+void core_mmu_get_user_pgdir(struct core_mmu_table_info *pgd_info)
+{
+	vaddr_t va_range_base = 0;
+	struct mmu_partition *prtn = core_mmu_get_prtn();
+	struct mmu_pgt *pgt = core_mmu_get_ta_pgt_va(prtn);
+
+	core_mmu_get_user_va_range(&va_range_base, NULL);
+	/*
+	 * In core_mmu_populate_user_map(), populates page table at level + 1
+	 * core_mmu_set_info_table(&pg_info, dir_info->level + 1, 0, NULL);
+	 Therefore, set tbl_info.level to -1.
+	 */
+	core_mmu_set_info_table(pgd_info, CORE_MMU_PGDIR_LEVEL - 1,
```

Comment:
The hard-code level in `core_mmu_populate_user_map()` has been fixed with
```
core_mmu_set_info_table(&pg_info, dir_info->next_level, 0, NULL);
```
Since `CORE_MMU_PGDIR_LEVEL` is 0, we might plus one here to get megapage level.
```suggestion
	core_mmu_set_info_table(pgd_info, CORE_MMU_PGDIR_LEVEL + 1,
```

Reply:
Agree.